File: MaWPhotos/src/main/java/us/mikeandwan/photos/ui/receiver/PhotoReceiverActivity.java
package us.mikeandwan.photos.ui.receiver;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.support.constraint.ConstraintLayout;
import android.support.design.widget.Snackbar;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.util.DisplayMetrics;
import android.util.Log;
import android.widget.TextView;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.InputStream;
import java.util.ArrayList;

import javax.inject.Inject;

import butterknife.BindDimen;
import butterknife.BindView;
import butterknife.ButterKnife;
import io.reactivex.Flowable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.schedulers.Schedulers;
import us.mikeandwan.photos.MawApplication;
import us.mikeandwan.photos.R;
import us.mikeandwan.photos.di.ActivityComponent;
import us.mikeandwan.photos.di.DaggerActivityComponent;
import us.mikeandwan.photos.services.DataServices;
import us.mikeandwan.photos.services.UploadJobScheduler;
import us.mikeandwan.photos.ui.BaseActivity;
import us.mikeandwan.photos.ui.HasComponent;


public class PhotoReceiverActivity extends BaseActivity implements HasComponent<ActivityComponent> {
    private final CompositeDisposable _disposables = new CompositeDisposable();
    private ActivityComponent _activityComponent;

    @Inject DataServices _dataServices;
    @Inject ReceiverRecyclerAdapter _receiverAdapter;
    @Inject UploadJobScheduler _uploadScheduler;

    @BindDimen(R.dimen.category_grid_thumbnail_size) int _thumbSize;
    @BindView(R.id.receiver_recycler_view) RecyclerView _recyclerView;
    @BindView(R.id.receiver_wifi_text_view) TextView _wifiTextView;
    @BindView(R.id.photoReceiverLayout) ConstraintLayout _layout;
    @BindView(R.id.toolbar) Toolbar _toolbar;


    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_photo_receiver);

        ButterKnife.bind(this);

        _activityComponent = DaggerActivityComponent.builder()
            .applicationComponent(getApplicationComponent())
            .activityModule(getActivityModule())
            .build();

        _activityComponent.inject(this);

        _recyclerView.setHasFixedSize(true);
        _recyclerView.setAdapter(_receiverAdapter);
        setLayoutManager();

        Intent intent = getIntent();
        String action = intent.getAction();

        if(action != null)
        {
            switch(action) {
                case Intent.ACTION_SEND:
                    handleSendSingle(intent);
                    break;
                case Intent.ACTION_SEND_MULTIPLE:
                    handleSendMultiple(intent);
                    break;
            }
        }

        // if we end up on this page, we either have new files to upload, or a user wants to check
        // so lets try to reschedule the job to kick it off
        _uploadScheduler.schedule(true);
    }


    @Override
    public void onResume() {
        updateToolbar(_toolbar, "Upload Queue");

        _disposables.add(_dataServices
            .getFileQueueObservable()
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe(
                this::updateListing
            ));

        super.onResume();
    }


    @Override
    protected void onDestroy() {
        _disposables.clear();

        super.onDestroy();
    }


    public ActivityComponent getComponent() {
        return _activityComponent;
    }


    private void setLayoutManager() {
        // https://stackoverflow.com/questions/33575731/gridlayoutmanager-how-to-auto-fit-columns
        DisplayMetrics displayMetrics = getResources().getDisplayMetrics();
        int cols = displayMetrics.widthPixels / _thumbSize;

        GridLayoutManager glm = new GridLayoutManager(this, Math.max(1, cols));
        _recyclerView.setLayoutManager(glm);
        _recyclerView.setItemAnimator(new DefaultItemAnimator());

        _receiverAdapter.setItemSize(displayMetrics.widthPixels / cols);

        _recyclerView.getRecycledViewPool().clear();
    }


    private boolean isValidType(String type) {
        return type != null && (type.startsWith("image/") || type.startsWith("video/"));
    }


    void handleSendSingle(Intent intent) {
        Uri mediaUri = intent.getParcelableExtra(Intent.EXTRA_STREAM);
        ArrayList<Uri> list = new ArrayList<>();

        list.add(mediaUri);

        saveFiles(list);
    }


    void handleSendMultiple(Intent intent) {
        ArrayList<Uri> mediaUris = intent.getParcelableArrayListExtra(Intent.EXTRA_STREAM);

        saveFiles(mediaUris);
    }


    void saveFiles(ArrayList<Uri> mediaUris) {
        _disposables.add(
            Flowable.fromCallable(() -> enqueueFiles(mediaUris))
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                    msg -> Snackbar.make(_layout, msg, Snackbar.LENGTH_LONG).show(),
                    ex -> {
                        Log.e(MawApplication.LOG_TAG, "error loading categories: " + ex.getMessage());
                        handleApiException(ex);
                    }
                ));
    }


    private String enqueueFiles(ArrayList<Uri> mediaUris) throws FileNotFoundException {
        int count = 0;
        int unsupportedFiles = 0;

        for(Uri uri : mediaUris) {
            String type = getContentResolver().getType(uri);

            if(!isValidType(type)) {
                unsupportedFiles++;
                continue;
            }

            InputStream inputStream = getContentResolver().openInputStream(uri);

            if(_dataServices.enequeFileToUpload(count + 1, inputStream, type))
            {
                count++;
            }
        }

        String msg = "";

        if(count > 0) {
            msg += "Successfully enqueued " + count + " files for upload.";
        }

        if(unsupportedFiles > 0) {
            if(msg.length() > 0) {
                msg += "  ";
            }

            msg += "Unable to enqueue " + unsupportedFiles + " files.";
        }

        return msg;
    }


    private void updateListing(File[] files) {
        _receiverAdapter.setQueuedFiles(files);
    }
}

File: MaWPhotos/src/main/java/us/mikeandwan/photos/ui/receiver/PhotoReceiverActivity.kt
package us.mikeandwan.photos.ui.receiver

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.GridLayoutManager
import com.google.android.material.snackbar.Snackbar
import dagger.hilt.android.AndroidEntryPoint
import io.reactivex.Flowable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import timber.log.Timber
import us.mikeandwan.photos.R
import us.mikeandwan.photos.databinding.ActivityPhotoReceiverBinding
import us.mikeandwan.photos.services.DataServices
import us.mikeandwan.photos.services.UploadJobScheduler
import us.mikeandwan.photos.ui.BaseActivity
import java.io.File
import java.io.FileNotFoundException
import java.util.*
import javax.inject.Inject

@AndroidEntryPoint
class PhotoReceiverActivity : BaseActivity() {
    private val _disposables = CompositeDisposable()
    private var _thumbSize = 0

    @Inject lateinit var _dataServices: DataServices
    @Inject lateinit var _receiverAdapter: ReceiverRecyclerAdapter
    @Inject lateinit var _uploadScheduler: UploadJobScheduler

    private lateinit var binding: ActivityPhotoReceiverBinding

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityPhotoReceiverBinding.inflate(layoutInflater)
        setContentView(binding.root)

        _thumbSize = resources.getDimension(R.dimen.image_grid_thumbnail_size_medium).toInt()
        binding.receiverRecyclerView.setHasFixedSize(true)
        binding.receiverRecyclerView.adapter = _receiverAdapter

        setLayoutManager()

        val intent = intent
        val action = intent.action

        if (action != null) {
            when (action) {
                Intent.ACTION_SEND -> handleSendSingle(intent)
                Intent.ACTION_SEND_MULTIPLE -> handleSendMultiple(intent)
            }
        }

        // if we end up on this page, we either have new files to upload, or a user wants to check
        // so lets try to reschedule the job to kick it off
        _uploadScheduler.schedule(true)
    }

    public override fun onResume() {
        updateToolbar(binding.toolbar, "Upload Queue")

        _disposables.add(_dataServices
            .fileQueueObservable
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe { files: Array<File>? -> updateListing(files!!) })

        super.onResume()
    }

    override fun onDestroy() {
        _disposables.clear()
        super.onDestroy()
    }

    private fun setLayoutManager() {
        // https://stackoverflow.com/questions/33575731/gridlayoutmanager-how-to-auto-fit-columns
        val displayMetrics = resources.displayMetrics
        val cols = displayMetrics.widthPixels / _thumbSize
        val glm = GridLayoutManager(this, Math.max(1, cols))
        binding.receiverRecyclerView.layoutManager = glm
        binding.receiverRecyclerView.itemAnimator = DefaultItemAnimator()
        _receiverAdapter.setItemSize(displayMetrics.widthPixels / cols)
        binding.receiverRecyclerView.recycledViewPool.clear()
    }

    private fun isValidType(type: String?): Boolean {
        return type != null && (type.startsWith("image/") || type.startsWith("video/"))
    }

    fun handleSendSingle(intent: Intent) {
        val mediaUri = intent.getParcelableExtra<Uri>(Intent.EXTRA_STREAM)
        val list = ArrayList<Uri?>()
        list.add(mediaUri)
        saveFiles(list)
    }

    fun handleSendMultiple(intent: Intent) {
        val mediaUris = intent.getParcelableArrayListExtra<Uri?>(Intent.EXTRA_STREAM)
        saveFiles(mediaUris)
    }

    fun saveFiles(mediaUris: ArrayList<Uri?>?) {
        _disposables.add(
            Flowable.fromCallable { enqueueFiles(mediaUris) }
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                    { msg: String? -> Snackbar.make(binding.photoReceiverLayout, msg!!, Snackbar.LENGTH_LONG).show() }
                ) { ex: Throwable ->
                    Timber.e("error loading categories: %s", ex.message)
                    handleApiException(ex)
                })
    }

    @Throws(FileNotFoundException::class)
    private fun enqueueFiles(mediaUris: ArrayList<Uri?>?): String {
        var count = 0
        var unsupportedFiles = 0
        for (uri in mediaUris!!) {
            val type = contentResolver.getType(uri!!)
            if (!isValidType(type)) {
                unsupportedFiles++
                continue
            }
            val inputStream = contentResolver.openInputStream(uri)
            if (_dataServices.enqueueFileToUpload(count + 1, inputStream!!, type!!)) {
                count++
            }
        }
        var msg = ""
        if (count > 0) {
            msg += "Successfully enqueued $count files for upload."
        }
        if (unsupportedFiles > 0) {
            if (msg.length > 0) {
                msg += "  "
            }
            msg += "Unable to enqueue $unsupportedFiles files."
        }
        return msg
    }

    private fun updateListing(files: Array<File>) {
        _receiverAdapter.setQueuedFiles(files)
    }
}